File: 210/level2/오픈채팅방.java
// 20220803

package level2;

import java.util.*;

class Main {
    static int N, M;
    static int[] Group;

    public static void main(String[] args) throws Exception {
        String s[] = { "Enter uid1234 Muzi", "Enter uid4567 Prodo", "Leave uid1234", "Enter uid1234 Prodo",
                "Change uid4567 Ryan", "Change uid4567 Ryatttt" };
        Solution sol = new Solution();

        System.out.println("result : " + sol.solution(s).length);
    }
}

class Solution {
    public String[] solution(String[] record) {
        List<Log> log = new ArrayList<>();

        HashMap<String,String> users = new HashMap<>();

        for (String records : record) {
            String[] currRecord = records.split(" ");

            if (currRecord[0].equals("Enter")) {
                log.add(new Log(currRecord[0], currRecord[1]));
                users.put(currRecord[1], currRecord[2]);
            } else if (currRecord[0].equals("Change")) {
                users.replace(currRecord[1], currRecord[2]);
            } else {
                log.add(new Log(currRecord[0], currRecord[1]));
            }
        }

        String[] answer = new String[log.size()];
        for (int i = 0; i < log.size(); i++) {
            if (log.get(i).state.equals("Enter")) {
                answer[i] = users.get(log.get(i).uid) + "님이 들어왔습니다.";
            } else {
                answer[i] = users.get(log.get(i).uid) + "님이 나갔습니다.";
            }
        }

        return answer;
    }

    class Log{
        String state;
        String uid;

        public Log(String state, String uid) {
            this.state = state;
            this.uid = uid;
        }
    }
}

File: Hprogram/level2/오픈채팅방.js
function solution(record) {
	var answer = [];
	let arr = [];
	let name = new Map();

	for (let el of record) {
		let split = el.split(" ");
		if (split[0] === "Enter") {
			arr.push([split[0], split[1]]); // ["Enter","uid1234"]
			name.set(split[1], split[2]); // 고유 id 기반으로 이름 맵핑. "uid1234" => "Muzi"
		} else if (split[0] === "Leave") {
			arr.push([split[0], split[1]]);
		} else if (split[0] === "Change") {
			name.set(split[1], split[2]);
		}
	}

	for (let el of arr) {
		let names = name.get(el[1]);
		if (el[0] === "Enter") {
			answer.push(`${names}님이 들어왔습니다.`);
		} else {
			answer.push(`${names}님이 나갔습니다.`);
		}
	}

	return answer;
}

// 삽입될 떄 마다 값을 배열을 사용해서 위치 저장
